gslib/utils/shim_util.py

```diff
@@ -0,0 +1,219 @@
+                self.command_name))
+      sub_command = gsutil_args[0]
+      sub_opts, parsed_args = self.ParseSubOpts(
+          args=gsutil_args[1:], should_update_sub_opts_and_args=False)
+      return self._get_gcloud_storage_args(
+          sub_opts, parsed_args,
+          gcloud_storage_map.gcloud_command.get(sub_command))
+    else:
+      raise ValueError('Incorrect mapping found for "{}" command'.format(
+          self.command_name))
+
+    if sub_opts:
+      for option, value in sub_opts:
+        if option not in gcloud_storage_map.flag_map:
+          raise exception.GcloudStorageTranslationError(
+              'Command option "{}" cannot be translated to'
+              ' gcloud storage'.format(option))
+        args.append(gcloud_storage_map.flag_map[option].gcloud_flag)
+        if value != '':
+          args.append(value)
+    return args + gsutil_args
+
+  def get_gcloud_storage_args(self):
+    """Translates the gsutil command flags to gcloud storage flags.
+
+    It uses the command_spec.gcloud_storage_map field that provides the
+    translation mapping for all the flags.
+    
+    Returns:
+      A list of all the options and arguments that can be used with the
+        equivalent gcloud storage command.
+    Raises:
+      GcloudStorageTranslationError: If a flag or command cannot be translated.
+      ValueError: If there is any issue with the mapping provided by
+        GcloudStorageMap.
+    """
+    return self._get_gcloud_storage_args(self.sub_opts, self.args,
+                                         self.gcloud_storage_map)
+
+  def _print_gcloud_storage_command_info(self,
+                                         gcloud_command,
+                                         env_variables,
+                                         dry_run=False):
+    logger_func = self.logger.info if dry_run else self.logger.debug
+    logger_func('Gcloud Storage Command: {}'.format(' '.join(gcloud_command)))
+    if env_variables:
+      logger_func('Enviornment variables for Gcloud Storage:')
+      for k, v in env_variables.items():
+        logger_func('%s=%s', k, v)
+
+  def translate_to_gcloud_storage_if_requested(self):
+    """Translates the gsutil command to gcloud storage equivalent.
+
+    The translated commands get stored at
+    self._translated_gcloud_storage_command.
+    This command also translate the boto config, which gets stored as a dict
+    at self._translated_env_variables
+    
+    Returns:
+      True if the command was successfully translated, else False.
+    """
+    use_gcloud_storage = config.get('GSUtil', 'use_gcloud_storage', 'never')
+    if use_gcloud_storage not in VALID_USE_GCLOUD_STORAGE_VALUES:
+      raise exception.CommandException(
+          'Invalid option specified for'
+          ' GSUtil:use_gcloud_storage config setting. Should be one of: {}'.
+          format(' | '.join(VALID_USE_GCLOUD_STORAGE_VALUES)))
+    if use_gcloud_storage != 'never':
```

Comment:
You can unindent everything below if you follow the "guard pattern."
```
if use_gcloud_storage == 'never'
  return
```

Also, ideally we use an enum instead of the raw string value "never"

Reply:
True, but used != to avoid repeating `return False` twice. 

Else we end up doing
```
if use_gcloud_storage == `never`:
  return False

# Perform bunch of other logic

return False.
```

To me, the way it is currently felt more intuitive. 

Regarding enum, I avoided it because it is python3 only, and for python2 we need to add another dependency which felt like overkill for a shim. I am instead using constants now which is slightly better than using raw stirngs directly.

Reply:
Gsutil doesn't support Python 2 anymore

Reply:
yes, but we still support backporting things in maintenance mode, and an invalid import will break all the tests. It's best to keep this until we have totally deprecated python 2.

Reply:
Aren't the backports done in a separate branch? https://github.com/GoogleCloudPlatform/gsutil/tree/py2

Reply:
Done